### ISSUE TITLE
[Question] RL Zoo uses dummy vec env by default?

### ISSUE THREAD

**Issue by zhihanyang2022** [DLR-RM/stable-baselines3#643]
### Question

I'm trying to understand how parallelism is implement is stable-baselines. 

By default, as in `train.py` in RL Zoo, it seems like `DummyVecEnv` is used because

- https://github.com/DLR-RM/rl-baselines3-zoo/blob/64d4625599d3244308f643810fd57d240aeeac58/train.py#L57
- From SB3's doc for PPO, to reproduce results in RL Zoo: `python train.py --algo ppo --env $ENV_ID --eval-episodes 10 --eval-freq 10000` (which doesn't change the default value for `vec-env` in the previous link)

However, by my understanding of the documentation, `DummyVecEnv` calls each env in sequence, and hence we shouldn't expect any speedup from it. 

Therefore, I have three quick questions:
1. Is there a reason why `SubprocVecEnv` is not used here? I mean, isn't it faster?
2. If `DummyVecEnv` was used, how long did training take (approximately) for 1 million timesteps for, let's say, `halfcheetah-v2`? Was the training duration reasonable?
3. Is `VecNormalize` used by default in both cases?

I have read through the documentation but couldn't find answers on these. Thanks for the help in advance.

### Checklist

- [x] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/) (**required**)
- [x] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo (**required**)

**Comment by Miffyli** [comment]
rl-baselines3-zoo questions should generally be opened in the respective repository :). 

I thought we had docs on this but even google could not find text this on this from us apart for Antonin's colab notebooks: https://colab.research.google.com/github/araffin/rl-tutorial-jnrr19/blob/master/3_multiprocessing.ipynb

But to give brief answers:

1) The overhead from multiprocessing (IPC etc.) may actually hurt performance, and especially in small environments using DummyVecEnv is much faster (I think same applies to MuJoCo environments).
2) I have no experience with MuJoCo so I can not pitch in here, but it should not take too long (I am guessing a hour or two at most).
3) You might need to enable this in hyperaparameters (much like done with the [pybullet variant](https://github.com/DLR-RM/rl-baselines3-zoo/blob/master/hyperparams/ppo.yml#L136-L159)). Normalization is not enabled by default.

**Comment by zhihanyang2022** [comment]
Thanks for comments~

1. Now I understand that overhead might be a problem for SubprocVenEnv and might make it worse than DummyVecEnv. But, for example, is DummyVecEnv of 10 envs faster than running 1 envs for 10 times longer? If so, why? I guess this confuses me the most.
3. I checked the documentation for such arguments, but didn't find any.

**Comment by Miffyli** [comment]
> Now I understand that overhead might be a problem for SubprocVenEnv and might make it worse than DummyVecEnv. But, for example, is DummyVecEnv of 10 envs faster than running 1 envs for 10 times longer? If so, why? I guess this confuses me the most.

One main contributor may be less calls to the agent `predict` function which is used to get actions during rollouts (we query actions for all environments with one `predict` call, regardless of the number of the environments). With more envs you better utilize the parallel nature of networks (bigger batches), where you have roughly the same runtime for predicting actions for a single environment or for eight. You will also end up with bigger batches of data before doing training, so there is less overhead between preparing rollout buffers and whatnot.

> I checked the documentation for such arguments, but didn't find any.

Ah sorry, I should have been more clear: I thought we had more documentation on this but it turns we do not :)

**Comment by zhihanyang2022** [comment]
> Ah sorry, I should have been more clear: I thought we had more documentation on this but it turns we do not :)

That's ok.

**Comment by zhihanyang2022** [comment]
My colleague was telling me about how PPO can be very fast due to parallelism over data collection. However, I don't think this is true, and I'm outlining my reasons below. Since I'm still learning PPO, please comment if you think they are incorrect :).

**First reason.** Given the same number of so-called "timesteps" to run, PPO takes drastically fewer total gradient steps than SAC. For example:

- 1 million steps of SAC means 1 million environment interactions and 1 million gradient steps
- 1 million steps of PPO means 1 million environment interactions and `1 million / (n_steps * n_envs) * n_epochs` gradient steps. Suppose `n_steps=100` and `n_envs=10` and `n_epochs=10`, this is only 10k gradient steps.  

Since gradient steps are in general more costly, taking fewer gradient steps certain reduce training time. Also, what's slightly disturbing is that increasing `n_envs` would also reduce number of gradient steps taken, provided that `n_steps` is unchanged. This seems to be **particular problematic** since SB3 offers no default value for `n_envs`.

**Second reason.** Both DummyVecEnv and SubprocVecEnv gives relatively poor speedup.

I constructed the following script to run 16000 steps:
- Across 16 envs, each 1000 steps, using SubprocVecEnv
- Across 16 envs, each 1000 steps, using DummyVecEnv
- Across 1 env, 16000 steps, without using VecEnv wrappers

My results are:
- SubprocVecEnv Duration:  4.417165219783783
- DummyVecEnv Duration:  0.9361192602664232
- Standard Duration:  1.500107979401946

which shows that speedup is Okayish but still relatively poor considering that we are parallelizing over 16 processes.

```python
import time
import numpy as np
import gym
from stable_baselines3.common.vec_env import SubprocVecEnv, DummyVecEnv


if __name__ == '__main__':

    example_env = gym.make("HalfCheetah-v2")

    # @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    # testing SubprocVecEnv


    def make_env():
        return gym.make("HalfCheetah-v2")


    env_fns = [make_env for _ in range(16)]
    env = SubprocVecEnv(env_fns)

    start = time.perf_counter()

    env.reset()
    for _ in range(1000):

        env.step_async(np.random.normal(size=(16, example_env.action_space.shape[0])))
        observation, reward, done, information = env.step_wait()

    end = time.perf_counter()

    duration = end - start

    print("SubprocVecEnv Duration: ", duration)

    # @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    # testing DummyVecEnv

    env_fns = [make_env for _ in range(16)]
    env = DummyVecEnv(env_fns)

    start = time.perf_counter()

    env.reset()
    for _ in range(1000):

        env.step_async(np.random.normal(size=(16, example_env.action_space.shape[0])))
        observation, reward, done, information = env.step_wait()

    end = time.perf_counter()

    duration = end - start

    print("DummyVecEnv Duration: ", duration)

    # @@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    # testing un-parallelized

    env = make_env()

    start = time.perf_counter()

    env.reset()
    for _ in range(16000):

        obs, reward, done, info = env.step(env.action_space.sample())

        if done:
            env = make_env()
            env.reset()

    end = time.perf_counter()

    duration = end - start

    print("Standard Duration: ", duration)
```

**Comment by araffin** [comment]
>  1 million environment interactions and 1 million / (n_steps * n_envs) * n_epochs gradient steps. Suppose n_steps=100 and n_envs=10 and n_epochs=10, this is only 10k gradient steps.

you are forgetting the batchsize... see https://github.com/DLR-RM/stable-baselines3/issues/632

>  Both DummyVecEnv and SubprocVecEnv gives relatively poor speedup

this highly depends on your environment (and on your hyperparameters).

**Comment by zhihanyang2022** [comment]
> you are forgetting the batchsize... see #632

Ooops. Thanks for pointing that out. Then the formula would be `1 million / (n_steps * n_envs) * n_epochs * (n_steps * n_envs) / batch_size = 1 million * n_epochs / batch_size`, which is confirmed by #632.

I think this confirms (at least for Mujoco envs where VecEnv doesn't speed things up that much) that a significant part of speedup by PPO compared to, e.g., SAC, comes from less gradient steps, since batch_size > n_epochs in general and the number of gradient steps would be much less than 1 million.

> this highly depends on your environment (and on your hyperparameters).

I'd really appreciate if you can go deeper into this, since I think most people don't actually test out all the options:
- What are some characteristics of an environment that affect the VecEnv choice? 
- What are some principles that tend to work well?

**Comment by araffin** [comment]
> What are some characteristics of an environment that affect the VecEnv choice?

it depends on where is the bottleneck. It is common that the `reset()` is the most costly operation (may take some seconds in robotic environments), in that case it is recommended to use sub-processes to do the reset in parallel.

the problem with sub-process is that it comes with a communication overhead (as shown in the colab linked in the documentation and by @Miffyli ) so for env that does require heavy computation like pendulum/cartpole, it does not make sense to use sub processes. However, you may observe a good speedup with atari games for instance (given that you have a good cpu).

Although, at some point, it also does not make sense to add more envs for two reasons (also discussed in the colab):
- adding more envs favor exploration and make the whole training less sample efficient
- the bottlebeck will comes from the gradient update (done sequentially and not asynchronously)

> this highly depends on your environment (and on your hyperparameters).

regarding the hyperparameters, I'm thinking of `n_epochs` and `train_freq / gradient_steps` which control a compromise between data collection and gradient update.
Again, if the time spent doing gradient update is much greater than the data collection, it does not really make sense to add more envs.
